File: static/solutions/java/src/main/java/com/epi/MergeContacts.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;

public class MergeContacts {
  // @include
  public static List<ContactList> mergeContactLists(
      List<ContactList> contacts) {
    return new ArrayList<>(new HashSet(contacts));
  }

  public static class ContactList {
    public List<String> names;

    ContactList(List<String> names) { this.names = names; }

    @Override
    public boolean equals(Object obj) {
      if (obj == null || !(obj instanceof ContactList)) {
        return false;
      }
      return this == obj
          ? true
          : new HashSet(names).equals(new HashSet(((ContactList)obj).names));
    }

    @Override
    public int hashCode() {
      return new HashSet(names).hashCode();
    }
  }
  //@exclude

  public static void main(String[] args) {
    List<ContactList> contacts
        = Arrays.asList(new ContactList(Arrays.asList("a", "b", "c")),
                        new ContactList(Arrays.asList("a", "c", "b")),
                        new ContactList(Arrays.asList("b", "c", "d")));
    List<ContactList> result = mergeContactLists(contacts);
    assert(result.size() == 2);
  }
}

File: static/solutions/merge_contacts.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <string>
#include <unordered_set>
#include <vector>

using std::hash;
using std::string;
using std::unordered_set;
using std::vector;

// @include
struct ContactList {
  // Equal function for hash.
  bool operator==(const ContactList& that) const {
    return unordered_set<string>(names.begin(), names.end()) ==
           unordered_set<string>(that.names.begin(), that.names.end());
  }

  vector<string> names;
};

// Hash function for ContactList.
struct HashContactList {
  size_t operator()(const ContactList& contacts) const {
    size_t hash_code = 0;
    for (const string& name : unordered_set<string>(contacts.names.begin(),
                                                    contacts.names.end())) {
      hash_code ^= hash<string>()(name);
    }
    return hash_code;
  }
};

vector<ContactList> MergeContactLists(const vector<ContactList>& contacts) {
  unordered_set<ContactList, HashContactList> unique_contacts(
      contacts.begin(), contacts.end());
  return {unique_contacts.begin(), unique_contacts.end()};
}
// @exclude

int main(int argc, char* argv[]) {
  vector<ContactList> contacts = {
      {{"a", "b", "c"}}, {{"a", "c", "b"}}, {{"b", "c", "d"}}};
  auto result = MergeContactLists(contacts);
  assert(result.size() == 2);
  return 0;
}
